samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,571 @@
+		       evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			printk("MQTT SUBACK error %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] SUBACK packet id: %u\n", __func__, __LINE__,
+		       evt->param.suback.message_id);
+		break;
+
+	default:
+		printk("[%s:%d] default: %d\n", __func__, __LINE__,
+		       evt->type);
+		break;
+	}
+}
+
+
+/**@brief Resolves the configured hostname and
+ * initializes the MQTT broker structure
+ */
+static void broker_init(void)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo *addr;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM
+	};
+
+	err = getaddrinfo(CONFIG_AWS_HOSTNAME, NULL, &hints, &result);
+	if (err) {
+		printk("ERROR: getaddrinfo failed %d\n", err);
+
+		return;
+	}
+
+	addr = result;
+	err = -ENOENT;
+
+	/* Look for address of the broker. */
+	while (addr != NULL) {
+		/* IPv4 Address. */
+		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+			struct sockaddr_in *broker4 =
+				((struct sockaddr_in *)&broker);
+
+			broker4->sin_addr.s_addr =
+				((struct sockaddr_in *)addr->ai_addr)
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(8883);
+			printk("IPv4 Address found 0x%08x\n",
+			       broker4->sin_addr.s_addr);
+			break;
+		} else {
+			printk("ai_addrlen = %u should be %u or %u\n",
+			       (unsigned int)addr->ai_addrlen,
+			       (unsigned int)sizeof(struct sockaddr_in),
+			       (unsigned int)sizeof(struct sockaddr_in6));
+		}
+
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+static int provision(void)
+{
+	int err;
+
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+					AmazonRootCA3_pem,
+					AmazonRootCA3_pem_len);
+	if (err < 0) {
+		printk("Failed to register ca certificate: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_PRIVATE_KEY,
+				 private_pem_key,
+				 private_pem_key_len);
+	if (err < 0) {
+		printk("Failed to register private key: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_SERVER_CERTIFICATE,
+				 public_pem,
+				 public_pem_len);
+	if (err < 0) {
+		printk("Failed to register public certificate: %d",
+		       err);
+		return err;
+	}
+
+	return 0;
+}
+
+static int provision_modem_key(void)
+{
+	static sec_tag_t sec_tag_list[] = {CONFIG_NRF_CLOUD_SEC_TAG};
+
+#if defined(CONFIG_BSD_LIBRARY)
+	{
+		int err;
+
+		/* Delete certificates */
+		nrf_sec_tag_t sec_tag = CONFIG_NRF_CLOUD_SEC_TAG;
+
+		for (nrf_key_mgnt_cred_type_t type = 0; type < 3; type++) {
+			err = nrf_inbuilt_key_delete(sec_tag, type);
+			printk("nrf_inbuilt_key_delete(%lu, %d) => result=%d",
+				sec_tag, type, err);
+		}
+
+		/* Provision CA Certificate. */
+		err = nrf_inbuilt_key_write(CONFIG_NRF_CLOUD_SEC_TAG,
```

Comment:
Why use nrf_inbuilt_key_write, not tls_credential_add?

Reply:
It's for the modem and not the tls implementation in the application.